onnxruntime/core/providers/cuda/reduction/reduction_functions.cu

```diff
@@ -0,0 +1,401 @@
+      value += v[i];
+    }
+  }
+
+#if __CUDA_ARCH__ >= 700
+  __syncwarp();
+#else
+  __syncthreads();
+#endif
+
+  // Warp-level reduction (storage change: register -> register).
+  // The values in a warp will be summed up to a scalar. After warp-level
+  // reduction, each block holds num_warps_in_block values in the shared memory.
+  TOut value_ = value;
+#pragma unroll
+  for (int stride = NUM_THREADS_PER_WARP / 2; stride > 0; stride /= 2) {
+    value_ += __shfl_down_sync(ALL_ONE_MASK, value_, stride);
+  }
+
+  // Return early if only one warp is used for reduction.
+  // Given a fixed amount of threads, we perfer threads over warps over blocks so that we never have cases such as
+  // 1. two blocks and each of them has only 1 warp (32 threads).
+  // 2. two warps and each of them has only 2 threads.
+  if (num_warps_in_block == 1) {
+    if (tid_in_grid == 0) {
+      // Compilation time if-else branch controlled by template argument can be
+      // optimized out, so there will be no branch in real computation phase.
+      if (DivideResultBySize) {
+        output[0] = TFinalOp()(value_ / TOut(size));
+      } else {
+        output[0] = TFinalOp()(value_);
+      }
+    }
+    return;
+  }
+
+  if (lid_in_block == 0) {
+    shared_memory[wid_in_block] = value_;
+  }
+
+  __syncthreads();
+
+  // Block-level reduction (storage change: shared memory -> global memory).
+  // The values in a block will be summed up to a scalar.
+  // Note that the values are stored in the shared memory.
+  // Here we assume that the size of shared_memory is smaller
+  // than num_warps_in_block, so we just keep halving the number
+  // of threads in each iteartion. Our assumption is always true because
+  // the size of shared_memory equals to the number of warps.
+#pragma unroll
+  for (int stride = NUM_WARPS_PER_BLOCK / 2; stride > 0; stride /= 2) {
+    if (tid_in_block + stride < num_warps_in_block) {
+      shared_memory[tid_in_block] += shared_memory[tid_in_block + stride];
+    }
+    __syncthreads();
+  }
+
+  // Return early if only one block is used for reduction.
+  if (num_blocks_in_grid == 1) {
+    if (tid_in_grid == 0) {
+      // Compilation time if-else branch controlled by template argument can be
+      // optimized out, so there will be no branch in real computation phase.
+      if (DivideResultBySize) {
+        output[0] = TFinalOp()(shared_memory[0] / TOut(size));
+      } else {
+        output[0] = TFinalOp()(shared_memory[0]);
+      }
+    }
+    return;
+  }
+
+  if (tid_in_block == 0) {
+    buffer[bid_in_grid] = shared_memory[0];
+  }
+
+  __threadfence();
+  __syncthreads();
+
+  // Grid-level reduciton. We use the last block to sum up values
+  // stored in the global buffer.
+  __shared__ bool is_last_block_done;
+
+  if (tid_in_block == 0) {
+    int* p_lock = reinterpret_cast<int*>(buffer + num_blocks_in_grid);
+    int count = atomicAdd(p_lock, 1);
+    is_last_block_done = (count == (num_blocks_in_grid - 1));
```

Comment:
atomic_add?

Reply:
No. It's for integer atomicAdd which is supported suince the very beginning of CUDA.